### ISSUE TITLE
Shift + Return deletes highlighted code in Help browser

### ISSUE THREAD

**Issue by unculcated** [supercollider/supercollider#3877]
Shift + Return doesn't execute. It erases text

**Comment by snappizz** [comment]
can you give SC version and OS? is the iinterpreter booted?

**Comment by patrickdupuis** [comment]
This happens when the interpreter isn't running. Highlighting text and pressing shift-return deletes the text.

**Comment by unculcated** [comment]
It is running. Latest compile. Qt 5.11.1
Thats all I could get.  It worked yesterday.

**Comment by unculcated** [comment]
Yes, Interpreter running. 

A grey box surrounds the text. This is all in the help browser

**Comment by unculcated** [comment]
Yes, Just a help Browser issue

**Comment by patrickdupuis** [comment]
Yes, I see this too.

**Comment by patrickdupuis** [comment]
Related issue #3777

**Comment by snappizz** [comment]
there are some known issues with help browser code eval. closing as likely duplicate of #3829 an #3777.

**Comment by scztt** [comment]
Re-opening - this is actually worthy of investigation - we can't capture Shift+Return in the IDE, IIRC we never receive the event. However, if someone was able to capture Shift+Return in the JavaScript code for the help docs, we could easily forward that back to the IDE. This is a little beyond my JavaScript depth, but I'd be happy to do the rest of the plumbing if someone can trigger an event when Shift+Return is pressed on the page.

**Comment by htor** [comment]
@scztt no problem emitting events from js, i can make pull for that. i've pushed https://github.com/supercollider/supercollider/pull/3904 right now, so if/when that gets merged i'll work on top of that pull.

**Comment by htor** [comment]
it looks like a good solution is to set up a [`QWebChannel`](https://doc.qt.io/qt-5/qwebchannel.html), catch the shift-enter keypress event in javascript and call a function in c++ whenever that happens. i've tried to implement [a standalone example](https://doc.qt.io/qt-5/qtwebchannel-standalone-example.html) which provides the bare minimum to set up the channel in `help_browser.cpp`. i'm stuck because i don't know c++ very well, but i'm sure someone around here who does know would easily do it. 

found out so far that you need to put `WebSockets` into `QtCollider/CMakeLists.txt` to get some of the code to compile. the rest i can't easily figure out. 

the js code for listening to shift-enter is simple:

```js
window.onkeydown = e => {
    if (e.shiftKey && e.key === 'Enter')
        // shift-enter is pressed
}
```

**Comment by brianlheim** [comment]
@htor - do you have your existing work in a branch on GitHub? If you want someone to help you out that would be a good way to start.

**Comment by htor** [comment]
@brianlheim yes, this is what i have so far: https://github.com/htor/supercollider/commit/309d1da4386ad44fa69defebcf7fa53185bb93ee

it won't build. i've put everything new in `help_browser.cpp`. the code i'm using is from https://github.com/liutaouni/QtExamples/tree/master/webchannel/standalone. any help would be appreciated just to get a web socket connection up and running!

**Comment by prko** [comment]
Hi,
in [60daca83] under macOS, CMD+Enter evaluates a code twice when evaluating any code(s) in help document.
Is it related to this issue? or is it related to #3881? or is it a new one?

**Comment by brianlheim** [comment]
@prko that's a separate issue, thanks for catching that. Mind opening a new ticket for it?

**Comment by brianlheim** [comment]
@htor Thanks for getting the ball rolling! I have a working fix for it (on macOS at least) here: https://github.com/brianlheim/supercollider/tree/topic/shift-enter-eval

I just need to clean it up a bit and then will be ready to submit a PR.

**Comment by htor** [comment]
nice! 

as you mentioned in slack this opens up new possibilities for grabbing information from the ide into the help browser (settings, config, etc.). as someone also mentioned the help browser is still needs to work without the ide, in cases where it's used outside separately (f.ex http://docs.sccode.org/) so some sort of progressive loading will be needed - maybe first apply defaults from js + css files, or if something cached use that, or if webchannel is available then fetch settings, colors, theme and apply them.

**Comment by brianlheim** [comment]
Closed in #3960.
